### PR TITLE
Add `closed` property to `Curve3D`

### LINKED ISSUE #6396: Add a `cyclic` property to `Curve2D` and `Curve3D` to make them closed automatically
Curve3D done as of https://github.com/godotengine/godot/pull/86195.

### Describe the project you are working on

Testing!

### Describe the problem or limitation you are having in your project

The current way the engine handles closed/cyclic curves makes them very hard to work with
This implementation also seems pretty hack-y, and is not how most other software do it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a `cyclic` property to `Curve2D` and `Curve3D`, that toggles if the curve is closed/cyclic. This is how most game engines and software handle it.

Could also be named `closed`

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Unlike now, the first and last points in the `Curve2D` and `Curve3D` point array would now have both In and Out positions. Maybe hide the unused handle when `cyclic` is off?

Setting `cyclic` to on makes the first and last point connected, now without making a new point to connect them. 
Attempting to add a point at the end of the curve in the path editor sets the spline to `cyclic` instead of adding a new point.
`Close Curve` (rename to `Toggle Cyclic`?) in the path editor would now simply toggle the property.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You could probably make a script that clamps the last point to the first, to make it easier to edit, but this proposal is about changing the node behavior to make it work like this out of the box.

### Is there a reason why this should be core and not an add-on in the asset library?

Improving the editor usability out of the box.

### LINKED ISSUE #8650: Add a Path3D editor Option to "mimic" a closed curve
### Describe the project you are working on

A plugin for procedural mesh generation.

### Describe the problem or limitation you are having in your project

It's no secret that in the past few years there's been a lot of discussion and proposals to implement "closed curves" when talking about Path3D/Path2D and Curve3D/Curve2D.
The limitation is clear: having a "closed" and "continous" curve is cumbersome and not easily achievable, but it's something you would want almost all times.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The implementation I have in mind is very simple and straightforward, should be easy to implement, and will not even need to modify the Curve3D class.
The idea is to just have an **option** in the Path3D editor that will "synchronize" the first and the last point of the curve **mimicking** a closed curve without any other hassle.
Basically when this option is _true_, the **last point** will have the **same position, handle position, and tilt** of the **first point**.
This "mode" could alse be automatically activated when pressing the "Close Curve" button, but this is not mandatory.
I think this also has a lot of pros with respect to implement a Curve3D is_closed property, that would require a lot of work and could even be undesirable at times.
I can probably try an implementation in the next days.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

path3d_screenshot

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This would be used a lot if someone wishes to work with paths and curves.

### Is there a reason why this should be core and not an add-on in the asset library?

This should be core behaviour and desirable to have directly in the engine.

### PR DESCRIPTION
**Pull request by GreenCrowDev**
Should close https://github.com/godotengine/godot-proposals/issues/6396 and close https://github.com/godotengine/godot-proposals/issues/8650

https://github.com/godotengine/godot/assets/62719360/27b8d2d2-4b51-4ea1-a221-8ea61fa51dbb

**TODO:**

- [x] Change color of start and end points (especially important when the curve is closed and you can't tell where's the start);
- [x] In and Out gizmo handles display, respectively for Start and End node, when `_closed = true`;
- [x] In and Out properties don't show up in inspector right now, but should be displayed when `_closed = true`;
- [x] ~Close Curve button could become a Closed Curve toggle, so that user can easily access the value through the already nice icon on the top bar of the viewport;~ Path3DEditorPlugin Close Curve button is now a "Toggle Open/Closed Curve" button that just toggles the curve `closed` property;
- [x] Documentation;
- [ ] Tests (can you suggest some? I'm not used to them);

**TODO (optional):**

- [ ] Draw indices text near the respective point to easily identify them (needs a bespoke solution as at the moment there's not a system in place to draw text near 3D gizmos);

If you have suggestions about what behaviour I should try to check if stuff is broken by this change let me know (for example something like `PathFollow3D`...).

### CODE REVIEW

File: scene/resources/curve.h [hunk 1]
```diff
@@ -263,6 +263,8 @@ class Curve3D : public Resource {
 	mutable Vector<size_t> points_in_cache;
 #endif
 
+	bool _closed = false;
```
Comment (AThousandShips) [inline scene/resources/curve.h:266]:
```suggestion
	bool closed = false;
```
This file doesn't apply `_` prefix to member variables, let's keep it uniform,

File: scene/resources/curve.cpp [hunk 2]
```diff
@@ -2321,6 +2377,8 @@ void Curve3D::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_get_data"), &Curve3D::_get_data);
 	ClassDB::bind_method(D_METHOD("_set_data", "data"), &Curve3D::_set_data);
 
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "_closed"), "set_closed", "is_closed");
```
Comment (AThousandShips) [inline scene/resources/curve.cpp:2380]:
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "closed"), "set_closed", "is_closed");
```
This should not be prefixed with `_` unless you intend it to be hidden in the documentation, which I don't think should be done here

File: scene/resources/curve.h (hunk 1 above)
Comment (matricola787) [reply scene/resources/curve.h:266]: Oh, okay, I see. I basically copied the PR for Line2D, but I guess I misunderstood how to use these, thank you 👍

**Comment by matricola787** [comment]
Should I commit the changes to the branch as they are ready or wait to complete all of them? I'm not sure if running CI/CD checks every time is a problem...
Right now I'm adding features to a _dev branch on top of this one.

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 3]
```diff
@@ -436,8 +443,42 @@ void Path3DGizmo::redraw() {
...
-		if (primary_handle_points.size()) {
-			add_handles(primary_handle_points, handles_material);
+		// Primary handles: manage different colors for first, intermediate and last points.
+		const int pc = primary_handle_points.size();
+		if (pc) {
+			// Need to define indices separately.
+			Vector<int> idx;
+			idx.resize(pc);
+			for (int j = 0; j < pc; j++) {
+				idx.write[j] = j;
```
Comment (AThousandShips) [inline editor/plugins/path_3d_editor_plugin.cpp:453]:
```suggestion
			int *idx_ptr = idx.ptrw();
			for (int j = 0; j < pc; j++) {
				idx_ptr[j] = j;
```
More efficient version, avoids COW

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 4]
```diff
@@ -436,8 +443,42 @@ void Path3DGizmo::redraw() {
...
+				add_handles(last_pt_handle_point, c->is_closed() ? handles_material : last_pt_handle_material, last_pt_id);
+			}
+
+			// Add handle for first point.
+			primary_handle_points.remove_at(0);
+			idx.remove_at(0);
+			add_handles(first_pt_handle_point, c->is_closed() ? closed_pt_handle_material : first_pt_handle_material, first_pt_id);
+
+			// Add handles for remaining intermediate points.
+			if (primary_handle_points.size()) {
```
Comment (AThousandShips) [inline editor/plugins/path_3d_editor_plugin.cpp:479]:
```suggestion
			if (!primary_handle_points.is_empty()) {
```

**Comment by GreenCrowDev** [comment]
Any idea when to merge this and set a milestone? 4.3 or 4.4?

**Comment by AThousandShips** [comment]
It hasn't been reviewed or decided on so it's in 4.x until some decision or review has been made 🙂

**Comment by Chubercik** [comment]
>Should I commit the changes to the branch as they are ready or wait to complete all of them? I'm not sure if running CI/CD checks every time is a problem...
>Right now I'm adding features to a _dev branch on top of this one.

No one ever reprimanded me for committing to a live PR branch, so I think it's fine - just be sure to squish your commits afterwards (see: [Pull request workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html)).

**Review by fire — APPROVED** [review]

**Review by Calinou — CHANGES_REQUESTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/29b3d9e9e538f0aa8effc8ad8bf19a2915292a89), I noticed some issues:

- The start/end color for gizmos only appears at certain camera angles. Perhaps two sprite gizmos are overlapping and only one of them is showing up?

https://github.com/godotengine/godot/assets/180032/8c311b77-6435-4136-9278-b3537c99c589

This also causes the colors to stop appearing depending on the number of points drawn:

https://github.com/godotengine/godot/assets/180032/b5a6d2a3-e152-4f8a-8c9b-53b1bb2c1cf8

They show up reliably in a top orthogonal view, which is what leads me to think it's an overlapping gizmo sorting issue:

image

Also, as an improvement, I suggest drawing the last line that goes from the last point to the first in a different color:

image

Mockup made in GIMP:

image

This way, you can tell which line is automatically being inferred by the engine as opposed as being designed by the user.

**Comment by GreenCrowDev** [comment]
Thanks for the review @Calinou!

Yes, this worked well before https://github.com/godotengine/godot/pull/87901, and there's this other PR https://github.com/godotengine/godot/pull/89847 that refines `Path3D` editor, ready to be merged.

Since they work on the same lines of code, I'd like to wait for https://github.com/godotengine/godot/pull/89847 to be merged, that defines stuff the lack of which makes difficult to finalize these last tweaks you mentioned.

Is that okay? 😊

**Comment by Chubercik** [comment]
> Since they work on the same lines of code, I'd like to wait for #89847 to be merged, that defines stuff the lack of which makes difficult to finalize these last tweaks you mentioned.

Let's go.

**Comment by GreenCrowDev** [comment]
I fixed the handles issue @Calinou.

Also fixed handles color in the curve edit mode, after the rebase.

I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR? The bones that describe the curve direction should be sufficient, in the meantime.

**Comment by GreenCrowDev** [comment]
I fixed the handles issue @Calinou.

Also fixed handles color in the curve edit mode, after the rebase.

I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR? The bones that describe the curve direction should be sufficient, in the meantime.

File: scene/resources/curve.cpp [hunk 5]
```diff
@@ -2274,13 +2340,13 @@ void Curve3D::_get_property_list(List<PropertyInfo> *p_list) const {
 		pi.usage &= ~PROPERTY_USAGE_STORAGE;
 		p_list->push_back(pi);
 
-		if (i != 0) {
+		if (i != 0 || is_closed()) {
```
Comment (AThousandShips) [inline scene/resources/curve.cpp:2343]:
```suggestion
		if (closed || i != 0) {
```

File: scene/resources/curve.cpp [hunk 6]
```diff
@@ -2274,13 +2340,13 @@ void Curve3D::_get_property_list(List<PropertyInfo> *p_list) const {
...
 			pi = PropertyInfo(Variant::VECTOR3, vformat("point_%d/in", i));
 			pi.usage &= ~PROPERTY_USAGE_STORAGE;
 			p_list->push_back(pi);
 		}
 
-		if (i != points.size() - 1) {
+		if (i != points.size() - 1 || is_closed()) {
```
Comment (AThousandShips) [inline scene/resources/curve.cpp:2349]:
```suggestion
		if (closed || i != points.size() - 1) {
```

**Comment by GreenCrowDev** [comment]
Changes implemented.

Also fixed a couple of bugs.

Not sure why it fails the check on Linux 🤔
```
gpg: keyserver receive failed: Connection reset by peer
Error: Process completed with exit code 2.
```

**Comment by AyyZerrAsa** [comment]
> I fixed the handles issue @Calinou.
> 
> Also fixed handles color in the curve edit mode, after the rebase.
> 
> I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR?

What do you mean in regards to not having access to control points? I made some changes in #90357 that caches baked distance at each original control point. Not sure whether that's relevant here. Great commit btw I was needing this, super suprised closed curves weren't a thing already.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 79de2eaae7fb3bb8f13031ac2ec9b19128b6f471), it works as expected. I think this is in a good state to be merged early in the 4.4 cycle.

https://github.com/godotengine/godot/assets/180032/2888f45a-0b1b-4999-a437-cdcbfa566f02

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 7]
```diff
@@ -281,8 +281,15 @@ void Path3DGizmo::redraw() {
...
 	Ref<StandardMaterial3D> path_tilt_muted_material = gizmo_plugin->get_material("path_tilt_muted_material", this);
 	Ref<StandardMaterial3D> handles_material = gizmo_plugin->get_material("handles");
+	Ref<StandardMaterial3D> first_pt_handle_material = gizmo_plugin->get_material("first_pt_handle");
+	Ref<StandardMaterial3D> last_pt_handle_material = gizmo_plugin->get_material("last_pt_handle");
+	Ref<StandardMaterial3D> closed_pt_handle_material = gizmo_plugin->get_material("closed_pt_handle");
 	Ref<StandardMaterial3D> sec_handles_material = gizmo_plugin->get_material("sec_handles");
 
+	first_pt_handle_material->set_albedo(Color(0.0, 1.0, 0.0));
+	last_pt_handle_material->set_albedo(Color(1.0, 0.0, 0.0));
+	closed_pt_handle_material->set_albedo(Color(1.0, 1.0, 0.0));
```
Comment (Calinou) [inline editor/plugins/path_3d_editor_plugin.cpp:291]: I find the pure yellow a bit hard to distinguish from white, maybe try a slightly orange color for the closed handle:

```suggestion
	first_pt_handle_material->set_albedo(Color(0.2, 1.0, 0.0));
	last_pt_handle_material->set_albedo(Color(1.0, 0.2, 0.0));
	closed_pt_handle_material->set_albedo(Color(1.0, 0.8, 0.0));
```

This also includes tweaks to the first/last point colors to look more visually pleasing.

image

image

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 8]
```diff
@@ -934,6 +973,14 @@ void Path3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 	Ref<Curve3D> curve = path->get_curve();
 
 	Ref<StandardMaterial3D> handle_material = get_material("handles", p_gizmo);
+	Ref<StandardMaterial3D> first_pt_handle_material = get_material("first_pt_handle", p_gizmo);
+	Ref<StandardMaterial3D> last_pt_handle_material = get_material("last_pt_handle", p_gizmo);
+	Ref<StandardMaterial3D> closed_pt_handle_material = get_material("closed_pt_handle", p_gizmo);
+
+	first_pt_handle_material->set_albedo(Color(0.0, 1.0, 0.0));
+	last_pt_handle_material->set_albedo(Color(1.0, 0.0, 0.0));
+	closed_pt_handle_material->set_albedo(Color(1.0, 1.0, 0.0));
```
Comment (Calinou) [inline editor/plugins/path_3d_editor_plugin.cpp:982]: Same as above:

```suggestion
	first_pt_handle_material->set_albedo(Color(0.2, 1.0, 0.0));
	last_pt_handle_material->set_albedo(Color(1.0, 0.2, 0.0));
	closed_pt_handle_material->set_albedo(Color(1.0, 0.8, 0.0));
```

**Review by Chubercik — APPROVED** [review]

**Comment by GreenCrowDev** [comment]
Thanks for the review!
I applied the suggestions and rebased :)

**Comment by GreenCrowDev** [comment]
Rebased :)

**Review by fire — APPROVED** [review]

**Comment by akien-mga** [comment]
This seems fine overall to me, but I question the usefulness of the `toggle_closed` method, which is equivalent to `closed = !closed`. Is this used often enough to use a dedicated method? I note that Line2D has a `closed` property but no `toggle_closed` method.

**Comment by GreenCrowDev** [comment]
> This seems fine overall to me, but I question the usefulness of the `toggle_closed` method, which is equivalent to `closed = !closed`. Is this used often enough to use a dedicated method? I note that Line2D has a `closed` property but no `toggle_closed` method.

I honestly don't recall why `toggle_closed` is there, I guess I'll remove it and force push again.

**Review by akien-mga — APPROVED** [review]

File: scene/resources/curve.cpp [hunk 9]
```diff
@@ -1596,13 +1609,16 @@ void Curve3D::_bake() const {
...
 
+		const int num_intervals = closed ? points.size() : points.size() - 1;
+
 #ifdef TOOLS_ENABLED
-		points_in_cache.resize(points.size());
+		points_in_cache.resize(closed ? (points.size() + 1) : points.size());
 		points_in_cache.set(0, 0);
 #endif
 
+		// Point Count: Begins at 1 to account for the last point
```
Comment (AThousandShips) [inline scene/resources/curve.cpp:1619]:
```suggestion
		// Point Count: Begins at 1 to account for the last point.
```

**Review by AThousandShips — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**